metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/CachedFieldNameConverterFactory.java

```diff
@@ -0,0 +1,155 @@
+ *
+ * <p>Defining a shorter expiration interval allows config changes to be recognized more
+ * quickly, but also can impact performance negatively.
+ */
+public class CachedFieldNameConverterFactory implements FieldNameConverterFactory {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  /**
+   * A cache that contains a {@link FieldNameConverter} for each sensor type.
+   *
+   * A user can alter the {@link FieldNameConverter} for a given sensor at any time
+   * by altering the Indexing configuration.  The actual {@link FieldNameConverter}
+   * in use for a given sensor will only change once the original converter has
+   * expired from the cache.
+   */
+  private Cache<String, FieldNameConverter> fieldNameConverters;
+
+  /**
+   * Creates a {@link CachedFieldNameConverterFactory}.
+   *
+   * @param expires The duration before {@link FieldNameConverter}s are expired.
+   * @param expiresUnits The units before {@link FieldNameConverter}s are expired.
+   */
+  public CachedFieldNameConverterFactory(int expires, TimeUnit expiresUnits) {
+
+    fieldNameConverters = createFieldNameConverterCache(expires, expiresUnits);
+  }
+
+  /**
+   * Creates a {@link CachedFieldNameConverterFactory} where the cache expires after 5 minutes.
+   */
+  public CachedFieldNameConverterFactory() {
+
+    this(5, TimeUnit.MINUTES);
+  }
+
+  /**
+   * Creates a {@link CachedFieldNameConverterFactory} using the given cache.  This should only
+   * be used for testing.
+   *
+   * @param fieldNameConverters A {@link Cache} containing {@link FieldNameConverter}s.
+   */
+  public CachedFieldNameConverterFactory(Cache<String, FieldNameConverter> fieldNameConverters) {
+
+    this.fieldNameConverters = fieldNameConverters;
+  }
+
+  /**
+   * Creates a cache of {@link FieldNameConverter}s, one for each source type.
+   *
+   * @return A cache of {@link FieldNameConverter}s.
+   */
+  private Cache<String, FieldNameConverter> createFieldNameConverterCache(int expire, TimeUnit expireUnits) {
+
+    return Caffeine
+            .newBuilder()
+            .expireAfterWrite(expire, expireUnits)
+            .build();
+  }
+
+  /**
+   * Create a new {@link FieldNameConverter}.
+   *
+   * @param sensorType The type of sensor.
+   * @param config The writer configuration.
+   * @return
+   */
+  @Override
+  public FieldNameConverter create(String sensorType, WriterConfiguration config) {
+
+    return fieldNameConverters.get(sensorType, (s) -> createInstance(sensorType, config));
+  }
+
+  /**
+   * Create a new {@link FieldNameConverter}.
+   *
+   * @param sensorType The type of sensor.
+   * @param config The writer configuration.
+   * @return
+   */
+  private FieldNameConverter createInstance(String sensorType, WriterConfiguration config) {
+
+    // default to the 'DEDOT' field name converter to maintain backwards compatibility
+    FieldNameConverter result = new DeDotFieldNameConverter();
```

Comment:
Might we want to call the enum when we create these objects?

```
FieldNameConverters.DEDOT.get();
```

Reply:
Sure, that makes sense.